libr/anal/fcn.c

```diff
@@ -69,6 +79,126 @@ R_API void r_anal_fcn_update_tinyrange_bbs(RAnalFunction *fcn) {
 	}
 }
 
+// _fcn_tree_{cmp_addr,calc_max_addr,free,probe} are used by interval tree.
+static int _fcn_tree_cmp_addr(const void *a_, const RBNode *b_) {
+	const RAnalFunction *a = (const RAnalFunction *)a_,
+		*b = container_of (b_, const RAnalFunction, rb);
+	ut64 from0 = a->addr, to0 = a->addr + a->_size,
+		from1 = b->addr, to1 = b->addr + b->_size;
+	if (from0 != from1) {
+		return from0 < from1 ? -1 : 1;
+	}
+	if (to0 != to1) {
+		return to0 - 1 < to1 - 1 ? -1 : 1;
+	}
+	return 0;
+}
+
+static void _fcn_tree_calc_max_addr(RBNode *node) {
+	RAnalFunction *fcn = FCN_CONTAINER (node), *fcn1;
+	int i;
+	fcn->rb_max_addr = fcn->addr + fcn->_size - 1;
+	for (i = 0; i < 2; i++) {
+		if (node->child[i]) {
+			fcn1 = container_of (node->child[i], RAnalFunction, rb);
+			if (fcn1->rb_max_addr > fcn->rb_max_addr) {
+				fcn->rb_max_addr = fcn1->rb_max_addr;
+			}
+		}
+	}
+}
+
+static void _fcn_tree_free(RBNode *node) {
+	// TODO after ownership transfered from fcns to fcn_tree
```

Comment:
Leak here

Reply:
The ownership is taken by `fcns` currently. Added more comments.